enhancements/installer/cloud-custom-dns.md

```diff
@@ -0,0 +1,658 @@
+
+### Non-Goals
+
+- We will not use the [Kubernetes External DNS provider](https://github.com/openshift/external-dns)
+and/or the [OpenShift External DNS Operator](https://github.com/openshift/external-dns-operator)
+to program dns entries across varied DNS providers. The External DNS Operator
+relies on being able to program the customer's DNS solution. Because 
+programmable DNS is precluded in some environments due to technical limitations
+or organizational policy, we cannot require programmable DNS.
+
+## Proposal
+
+For the OpenShift installation to make use of a custom DNS solution that is not
+configurable by OpenShift, the customer is responsible for configuring the
+DNS entries for `api` and `*.apps` services.
+
+When it is indicated to the OpenShift installer via an input parameter that the
+custom DNS solution is in use, OpenShift does not configure the respective cloud
+DNS solutions [Route53 for AWS](https://aws.amazon.com/route53/), [Azure DNS](https://learn.microsoft.com/en-us/azure/dns/dns-overview) and [Cloud DNS for GCP](https://cloud.google.com/dns).
+
+The Installer configures the LBs for the API and API-Int services and the
+Ingress Controllers configure the LBs for the *.apps service. There is
+currently no way of knowing these LB IP addresses before their creation. So, 
+the customer would have to wait to configure their custom DNS solution until
+after the LBs are created by OpenShift and the cluster installation has
+completed.
+
+For the cluster installation to succeed before the custom DNS solution is setup
+for `api`, `api-int` and `*.apps` resolution, OpenShift will have to provide a
+self-hosted in-cluster DNS solution. This would allow the customer to configure
+their DNS solution post-install. The current proposal focuses on this approach.
+
+OpenShift will start static CoreDNS pods to provide DNS resolution for API,
+Internal API and Ingress services that are essential for cluster creation.
+In order for a master or worker node to come online, they need the
+machine config server via the api-int domain. So, it is essential for the
+api-int domain to be resolvable on the bootstrap node. The static CoreDNS
+pod started on the bootstrap node needs entries for just api-int resolution
+and the static CoreDNS pod started on the control plane nodes needs to be
+able to resolve api, api-int and *.apps domains. Although not identical, this
+approach leverages learnings from OpenShift's approach to providing in-cluster 
+DNS for on-prem platforms.
+
+After cluster deployment is completed, the customer will update their external
+DNS solution with the same assigned LB IP addresses used for the configuration
+of the internal CoreDNS instance. OpenShift will not delete the CoreDNS pod
+even after cluster installation completes.
+
+If the user successfully configures their external DNS service with api,
+api-int and *.apps services, then they could optionally delete the in-cluster
```

Comment:
The CoreDNS pod would be manged by the MCO. Currently there is no way of knowing if the lb addresses have changed since the Installer first created it. So, the lb dns addresses within the CoreDNS pod are also not expected to change. 
A new field is being added to `platformSpec` for AWS, Azure and GCP which indicates whether the customDNS solution has been enabled. The coreDNS pod would be created by the MCO only when the feature is `Enabled` and the ConfigMap containing the LB config is present. When either of these conditions becomes `False`, the CoreDNS pod could be deleted. These are manual steps. 
We are not recommending that the customer delete the coreDNS pod but pointing out that if the customer's DNS solution is configured correctly, then the cluster could function without the self-hosted coreDNS.

Reply:
Yes, I think we should make sure the user provisioned DNS is functioning before we remove the CoreDNS pod. So, if this capability is disabled day-2, the responsibility to check for a functioning DNS alternative would fall on which component? Or do we assume that the customer knows what they are doing and not have any checks? 
As you can tell, I have some unanswered questions in this area. I think if we decide to allow that in the future, we should be able to add something to the `Spec` to control that behavior.

Reply:
If the customer disables the feature, and we remove the pod without any checks, what will break? And how hard would it be for them to then recover the cluster?

Reply:
> If the customer disables the feature, and we remove the pod without any checks, what will break? And how hard would it be for them to then recover the cluster?

IMO, disabling the feature would mean that the customer would want to start using the cloud default DNS and discontinue using their external DNS and the in-cluster DNS. When the feature is disabled, the customer could remove/delete entries from their external DNS. MCO could delete the CoreDNS pod. How do we configure the cloud default DNS then? We could:
1. Ask the customer to manually configure the cloud DNS using LB values gathered from Infra CR or cloud CLI
2. The Installer is currently responsible for configuring the cloud DNS for API and API-Int. If the cloud DNS has to be configured with these values day-2, MCO (or another appropriate component has to take on this task).

My current understanding is that disabling this feature day-2 would be an exception. Providing just the manual option seems sufficient at this time. @JoelSpeed, @zaneb